fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricCodecDataProvider.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.datagen.v1.provider;
+
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+
+import com.google.gson.JsonElement;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.JsonOps;
+
+import net.minecraft.data.DataOutput;
+import net.minecraft.data.DataProvider;
+import net.minecraft.data.DataWriter;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.datagen.v1.FabricDataGenerator;
+import net.fabricmc.fabric.api.datagen.v1.FabricDataOutput;
+
+/**
+ * Extend this class and implement {@link FabricCodecDataProvider#configure}.
+ *
+ * <p>Register an instance of the class with {@link FabricDataGenerator.Pack#addProvider} in a {@link net.fabricmc.fabric.api.datagen.v1.DataGeneratorEntrypoint}.
+ */
+public abstract class FabricCodecDataProvider<T> implements DataProvider {
+	private final DataOutput.PathResolver pathResolver;
+	private final Codec<T> codec;
+
+	protected FabricCodecDataProvider(FabricDataOutput dataOutput, DataOutput.OutputType outputType, String directoryName, Codec<T> codec) {
+		this.pathResolver = dataOutput.getResolver(outputType, directoryName);
+		this.codec = codec;
+	}
+
+	@Override
+	public CompletableFuture<?> run(DataWriter writer) {
+		Map<Identifier, JsonElement> entries = new HashMap<>();
+		BiConsumer<Identifier, T> provider = (id, value) -> {
+			JsonElement json = this.convert(id, value);
+			JsonElement existingJson = entries.put(id, json);
+
+			if (existingJson != null) {
+				throw new IllegalArgumentException("Duplicate entry " + id);
+			}
+		};
+
+		this.configure(provider);
+		return this.write(writer, entries);
+	}
+
+	protected abstract void configure(BiConsumer<Identifier, T> provider);
```

Comment:
This method could use a short javadoc comment like in the related methods in other providers

Reply:
True, then it would probably be useful to create another PR that adds javadoc to some of the other methods like [`FabricDynamicRegistryProvider::configure`](https://github.com/FabricMC/fabric/blob/1.19.4/fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricDynamicRegistryProvider.java#L71)?